Enable adjustment/selection of best virtual fit result

sadhana-r opened the pull request:
Closes #357 

- In the pre-planning module, the user is now able to see all of the computed virtual fit results.
- Multiple virtual fit results can be approved for a target
- The user can manually edit the transforms
- The user can `Add` a new transform node and manually perform virtual fit

A big change is that I have now removed the target option from the transducer tracking module. The target and virtual fit used for tracking is now based on what is selected in the Pre-planning module. I have added a `setVirtualFitResultForTracking` function to the transducer tracking module, which updates a `self._virtual_fit_transform_for_tracking` variable. This keeps track of the currently selected virtual fit result, and is used by the wizard, distance calculator, display etc.. This variable depends on the currently selected target and virtual result in the pre-planning module and the function is triggered by the pre-planning module everytime a different virtual fit result is selected in the table.

### For Review

This ended up being quite a massive change to the virtual fit workflow and how it interacts with the transducer tracking module. Please play around with different sessions and virtual fit/tracking interactions and note any bugs/weird behavior.

sadhana-r commented on OpenLIFUPrePlanning/OpenLIFUPrePlanning.py:
```diff
@@ -574,40 +635,142 @@ def setVirtualFitProgressDisplay(self, value: int, status_text: str):
+        selected_item = self.ui.virtualFitResultTable.findItems(node.GetName(), qt.Qt.MatchExactly)
+        if not selected_item:
+            raise RuntimeError("Cannot find the given node in the virtual fit results table")
+        self.ui.virtualFitResultTable.selectRow(selected_item[0].row())
```
Nice, thanks!

sadhana-r commented on OpenLIFUPrePlanning/Resources/UI/OpenLIFUPrePlanning.ui:
```diff
@@ -280,6 +273,50 @@
+             </widget>
+            </item>
+            <item>
+             <widget class="QTableWidget" name="virtualFitResultTable"/>
```
good catch! I disabled editing

arhowe00 commented on OpenLIFUTransducerTracker/OpenLIFUTransducerTracker.py:
```diff
@@ -838,9 +838,22 @@ def onTransformModified(self, node, eventID,):
+            self.transducer_to_volume_transform_node = transducer_transform_node_from_openlifu(
+                openlifu_transform_matrix = np.eye(4) ,
+                transducer = transducer.transducer.transducer,
+                transform_units = transducer.transducer.transducer.units)
```
The initial `transducer` doesn't look like it's declared anywhere

arhowe00 commented on OpenLIFUPrePlanning/OpenLIFUPrePlanning.py:
```diff
@@ -574,40 +636,143 @@ def setVirtualFitProgressDisplay(self, value: int, status_text: str):
+            approved : bool = get_approval_from_virtual_fit_result_node(vf_results[0])
+            self.updateApproveButton(approved)
+
+        self.ui.virtualFitResultTable.resizeColumnsToContents()
+        self.ui.virtualFitResultTable.resizeRowsToContents()
+        self.ui.virtualFitResultTable.setSelectionBehavior(qt.QAbstractItemView.SelectRows)
+        self.ui.virtualFitResultTable.setSelectionMode(qt.QAbstractItemView.SingleSelection) # Allow single row selection
+        self.ui.virtualFitResultTable.horizontalHeader().setSectionResizeMode(0, qt.QHeaderView.Stretch)
+
```
```suggestion
```

Same thing here. I removed these because I just specified these things in the ui file. I even removed the stuff about resizing and stretching because it's unlikely any row entry would significantly modify the table such that the stretch and resize would be necessary to display the content. Overall, the ui stuff is less important but I'm going for simplicity of changes. If you'd prefer this method, then reject the changes in the ui and keep these .

arhowe00 commented on OpenLIFUPrePlanning/Resources/UI/OpenLIFUPrePlanning.ui:
```diff
@@ -280,6 +273,54 @@
+             <widget class="QTableWidget" name="virtualFitResultTable">
+              <property name="editTriggers">
+               <set>QAbstractItemView::NoEditTriggers</set>
+              </property>
```
```suggestion
              </property>
              <property name="selectionMode">
               <enum>QAbstractItemView::SingleSelection</enum>
              </property>
              <property name="selectionBehavior">
               <enum>QAbstractItemView::SelectRows</enum>
              </property>
              <property name="columnCount">
               <number>2</number>
              </property>
              <attribute name="horizontalHeaderDefaultSectionSize">
               <number>200</number>
              </attribute>
              <attribute name="horizontalHeaderStretchLastSection">
               <bool>true</bool>
              </attribute>
              <column>
               <property name="text">
                <string>Name</string>
               </property>
              </column>
              <column>
               <property name="text">
                <string>Approved</string>
               </property>
              </column>
```

These are all the changes expressed in the `ui` file.

arhowe00 commented on OpenLIFUTransducerTracker/OpenLIFUTransducerTracker.py:
```diff
@@ -2091,35 +2117,71 @@ def updateApprovalStatusLabel(self):
-    def updateVirtualFitResultForDisplay(self):
+        if self._running_wizard:
+            return
+        current_result = self._virtual_fit_transform_for_tracking
```
```suggestion
```
(dead code)

arhowe00 reviewed (changes requested):
A few removals of code and one important static type change

arhowe00 reviewed (approved):
Amazing work on the changes! I'm going to approve because the workflow works without any immediate issues and the code has been reviewed and looked over a few times.  A few things to consider for the future:

1. It would be nice to officially rename virtual fit results, I can imagine a scenario where people are deliberating over a few options and maybe want to reference them through some meaningful label.
2. I can see that running the auto-fitting algorithm resets all VF results. Maybe we should warn (and possibly even ask how many to add, but these are all just sliders and switches that interact with what's already possible)
3. Aside from the auto scrolling on new VF results, there seems to be some weird scrolling behavior when clicking on VF results.

sadhana-r commented:
> 1. It would be nice to officially rename virtual fit results, I can imagine a scenario where people are deliberating over a few options and maybe want to reference them through some meaningful label.

Agree. We plan to do this in the future when we refine the virtual fit UI. This requires some changes to openlifu-python to be able to save the virtual fit result labels as part of the session.

> 2. I can see that running the auto-fitting algorithm resets all VF results. Maybe we should warn (and possibly even ask how many to add, but these are all just sliders and switches that interact with what's already possible)

Good point. For now, I added a sentence to the tooltip to indicate that current results are cleared. And also added a `notification`. Ideally, we should only clear the previously computed automated onces. This requires a larger changes to how we store and reference the virtual fit result nodes. I think we could add a 'Manual' vs 'Automatic' attribute to the transform_node to only clear the automated ones. Currently, when loading a session, we have no way of knowing which transforms were created manually vs using the algorith. So we can implement this change once we update openlifu.

> 3. Aside from the auto scrolling on new VF results, there seems to be some weird scrolling behavior when clicking on VF results.

I removed the setting to center the scrollbar after selection. It seemed to be causing the weird behaviour.
